src/test/java/part2/exercise/CollectorsExercise2.java

```diff
@@ -163,95 +163,105 @@ public MapPair(Map<String, Key> keyById, Map<String, List<Value>> valueById) {
+        }
+
+        @Override
+        public BiConsumer<MapPair, Pair> accumulator() {
+            return (m, p) -> {
+                final Key key = p.getKey();
+                final String sKey = key.toString();
+                final Value value = p.getValue();
+                final Map<String, Key> keyMap = m.getKeyById();
+                final Map<String, List<Value>> valueMap = m.getValueById();
+                keyMap.put(sKey, key);
+                if (!valueMap.containsKey(sKey))
+                    valueMap.put(sKey, new ArrayList<>());
+                valueMap.get(sKey).add(value);
+            };
+        }
+
+        @Override
+        public BinaryOperator<MapPair> combiner() {
+            return (a, b) -> {
+                final BinaryOperator<Map<String, Key>> keyMerger = mapMerger((k1, k2) -> k1);
+                final BinaryOperator<Map<String, List<Value>>> valueMerger = mapMerger((v1, v2) -> { v1.addAll(v2); return v1; });
+                return new MapPair(
+                        keyMerger.apply(a.getKeyById(), b.getKeyById()),
+                        valueMerger.apply(a.getValueById(), b.getValueById()));
+            };
+        }
+
+        @Override
+        public Function<MapPair, MapPair> finisher() {
+            return Function.identity();
+        }
+
+        @Override
+        public Set<Characteristics> characteristics() {
+            return Collections.unmodifiableSet(EnumSet.of(
+                    Characteristics.UNORDERED,
+                    Characteristics.IDENTITY_FINISH));
+        }
+
+    }
+
+    public static Map<Key, List<Value>> collectKeyValueMapPair(List<Pair> pairs) {
+
+        final MapPair mapPair = pairs.stream().collect(new MapPairCollector());
+        final Map<String, Key> keyMap = mapPair.getKeyById();
+        final Map<String, List<Value>> valuesMap = mapPair.getValueById();
+
+        return valuesMap.entrySet().stream()
+                .collect(
+                        toMap(e -> keyMap.get(e.getKey()), Map.Entry::getValue));
+
+    }
+
     @Test
     public void collectKeyValueMap() {
         final List<Pair> pairs = generatePairs(10, 100);
+        final Map<Key, List<Value>> expectedWithoutStream = collectKeyValueMapWithoutStream(pairs);
+
+        assertThat(collectKeyValueMapDoubleIteration(pairs).entrySet(), everyItem(isIn(expectedWithoutStream.entrySet())));
+        assertThat(collectKeyValueMapPair(pairs).entrySet(), everyItem(isIn(expectedWithoutStream.entrySet())));
 
-        // В два прохода
-        // final Map<String, Key> keyMap1 = pairs.stream()...
-
-        // final Map<String, List<Value>> valuesMap1 = pairs.stream()...
-
-        // В каждом Map.Entry id ключа должно совпадать с keyId для каждого значения в списке
-        // final Map<Key, List<Value>> keyValuesMap1 = valueMap1.entrySet().stream()...
-
-        // В 1 проход в 2 Map с использованием MapPair и mapMerger
-        final MapPair res2 = pairs.stream()

```

Comment:
There is MapPairCollector class and it's implementation in function collectKeyValueMapPair